### PR TITLE
feat(torch): torch detection training

### PR DESCRIPTION
**Pull request by Bycob**
Support of Faster RCNN & RetinaNet training on CPU at batch size 1
only.

Training on GPU leads to a segfault: https://github.com/pytorch/vision/issues/3525
EDIT: Fixed by upgrading pytorch to 1.8, see #1222

Models must be traced with a modified torchvision until this bug is resolved: 
https://github.com/pytorch/vision/issues/3469
`trace_torchvision` script uses the same torchvision used to build dede, which has the necessary modification.

TODO
- [x] working Faster RCNN training
- [ ] working RetinaNet training
- [x] Check that the MAP returned by dede is correct

```bash
# trace model with 2 classes
python3 tools/torch/trace_torchvision.py -v fasterrcnn_resnet50_fpn -o "/opt/platform/models/training/fasterrcnn_test_coco/" --num_classes 2

# service creation call
curl -X PUT 'http://localhost:8080/services/fasterrcnn_train' -d '{
    "description": "Torch traced fasterrcnn on cars",
    "mllib": "torch",
    "model": {
        "repository": "/opt/platform/models/training/fasterrcnn_train_cars/"
    },
    "parameters": {
        "input": {
            "connector": "image",
            "width":224,
            "height":224,
            "rgb":true,
            "scale":0.0039,
            "bbox":true
        },
        "mllib":{
            "template":"fasterrcnn",
            "nclasses":2,
            "gpu":true,
            "gpuid":0
        },
        "output": {
            "bbox": true,
            "store_config": true
        }
    },
    "mltype": "detection",
    "type": "supervised"
}'

# train call
curl -X POST 'http://localhost:8080/train' -d '{
   "service":"fasterrcnn_train",
   "parameters":{
      "input":{
         "db":true
      },
      "output":{
         "measure":[
            "map"
         ]
      },
      "mllib":{
         "solver":{
            "iterations":5000,
            "base_lr":1e-5,
            "finetune":true,
            "iter_size":16,
            "solver_type":"ADAM",
            "test_interval":250
         },
         "net":{
            "batch_size":1
         },
         "nclasses":1,
         "resume":false
      }
   },
   "data":[
      "/opt/platform/examples/cars/train.txt",
      "/opt/platform/examples/cars/test.txt"
   ]
}'
```

### CODE REVIEW

File: src/backends/torch/torchinputconns.cc [hunk 1]
```diff
@@ -351,6 +396,56 @@ namespace dd
...
+                tests_lfiles.resize(_uris.size() - 1);
+                for (size_t i = 1; i < _uris.size(); ++i)
+                  {
+                    read_image_bbox(tests_lfiles[i - 1], _uris.at(i));
+                  }
+
+                if (_dataset._shuffle)
+                  shuffle_dataset<std::string>(lfiles);
+
+                // bool has_test_data = test_lfiles.size() != 0;
```
Comment (fantes) [inline src/backends/torch/torchinputconns.cc:416]: remove this ?

File: src/backends/torch/torchlib.cc [hunk 2]
```diff
@@ -1530,13 +1578,165 @@ namespace dd
...
+      }
     ad_res.add("batch_size",
                entry_id); // here batch_size = tested entries count
     SupervisedOutput::measure(ad_res, ad_out, out, test_id, test_name);
     _module.train();
     return 0;
   }
 
+  double area(const std::vector<double> &bbox)
+  {
```
Comment (fantes) [inline src/backends/torch/torchlib.cc:1595]: maybe move below functions out of torchlib.cc?

**Review by fantes — APPROVED** [review]
nice work !

File: src/backends/torch/torchlib.cc (hunk 2 above)
Comment (Bycob) [reply src/backends/torch/torchlib.cc:1595]: Do you have a suggestion for where it could be moved? I think these function can apply to all backends, so maybe in `src/utils`?

Comment (fantes) [reply src/backends/torch/torchlib.cc:1595]: ok for src/utils :)

**Comment by beniz** [comment]
@Bycob can we specify the `loss` automatically instead of having to set it to `model` ?

File: src/backends/torch/torchdataset.cc [hunk 3]
```diff
@@ -525,6 +525,75 @@ namespace dd
...
+
+    while (std::getline(infile, line))
+      {
+        std::istringstream iss(line);
+        std::string val;
+        iss >> val;
+        int cls = std::stoi(val);
+        classes.push_back(target_to_tensor(cls));
+
+        std::vector<double> bbox;
```
Comment (beniz) [inline src/backends/torch/torchdataset.cc:577]: you may want to reserve the size of the vector since it is known in advance.

File: src/backends/torch/torchlib.cc [hunk 4]
```diff
@@ -725,7 +731,13 @@ namespace dd
                                              + e.what());
               }
 
-            Tensor loss = tloss.loss(y_pred, y, in_vals);
+            Tensor loss;
+            if (_loss == "model")
```
Comment (beniz) [inline src/backends/torch/torchlib.cc:735]: Since the template is known, and if this is the only correct loss option for these object detection models, I believe we should set it automatically.

File: src/backends/torch/torchlib.cc [hunk 5]
```diff
@@ -1150,7 +1162,7 @@ namespace dd
 
             if (bbox)
               {
-                // Supporting only Faster RCNN format at the moment.
+                // Supporting only torchvision format at the moment.
```
Comment (beniz) [inline src/backends/torch/torchlib.cc:1165]: you mean `torchvision output format` ?

File: src/backends/torch/torchlib.cc (hunk 2 above)
Comment (beniz) [reply src/backends/torch/torchlib.cc:1595]: you can also create `src/bbox_utils` files, it is a common practice as bboxes require some specific ops.

File: src/backends/torch/torchmodule.cc [hunk 6]
```diff
@@ -294,7 +294,16 @@ namespace dd
             source = torch_utils::unwrap_c10_vector(output);
           }
       }
-    c10::IValue out_val = source.at(_linear_in);
+    c10::IValue out_val;
+    if (_training && _loss_id >= 0)
```
Comment (beniz) [inline src/backends/torch/torchmodule.cc:298]: Maybe add a comment that describes what this block does and how loss_id works.

File: src/backends/torch/torchmodule.h [hunk 7]
```diff
@@ -195,7 +195,10 @@ namespace dd
     torch::nn::Linear _linear = nullptr;
 
     torch::Device _device;
-    int _linear_in = 0; /**<id of the input of the final linear layer */
+    bool _training = false; /**<true if model is in training mode */
+    int _linear_in = 0;     /**<id of the input of the final linear layer */
```
Comment (beniz) [inline src/backends/torch/torchmodule.h:199]: does this only apply to the object detection models from torchvision ?

File: src/backends/torch/torchlib.cc (hunk 2 above)
Comment (Bycob) [reply src/backends/torch/torchlib.cc:1595]: So `src/utils/bbox.hpp`?

File: src/backends/torch/torchlib.cc (hunk 5 above)
Comment (Bycob) [reply src/backends/torch/torchlib.cc:1165]: yes

File: src/backends/torch/torchmodule.h (hunk 7 above)
Comment (Bycob) [reply src/backends/torch/torchmodule.h:199]: `_training` field is set when we call the method train(). Currently it's used only for models that have different output on training and eval mode, ie detection models.

File: src/utils/bbox.hpp [hunk 8]
```diff
@@ -0,0 +1,66 @@
+/**
+ * DeepDetect
+ * Copyright (c) 2014-2015 Emmanuel Benazera
```
Comment (beniz) [inline src/utils/bbox.hpp:3]: I don't remember writing this :) I let you update the copyright and email!

Comment (Bycob) [reply src/utils/bbox.hpp:3]: Sure :)

**Review by beniz — APPROVED** [review]
